samples/net/mqtt_publisher/src/main.c

```diff
@@ -294,111 +325,86 @@ static int try_to_connect(struct mqtt_client_ctx *client_ctx)
-	client_ctx.mqtt_ctx.entropy_src_cb = NULL;
+#if defined(CONFIG_MQTT_LIB_TLS)
+	rc = tls_init();
+	PRINT_RESULT("tls_init", rc);
+	if (rc < 0) {
+		return;
+	}
 #endif
 
-	/* Publisher apps TX the MQTT PUBLISH msg */
-	client_ctx.mqtt_ctx.publish_tx = publish_cb;
-
-	/* The connect message will be sent to the MQTT server (broker).
-	 * If clean_session here is 0, the mqtt_ctx clean_session variable
-	 * will be set to 0 also. Please don't do that, set always to 1.
-	 * Clean session = 0 is not yet supported.
-	 */
-	client_ctx.connect_msg.client_id = MQTT_CLIENTID;
-	client_ctx.connect_msg.client_id_len = strlen(MQTT_CLIENTID);
-	client_ctx.connect_msg.clean_session = 1;
-
-	client_ctx.connect_data = "CONNECTED";
-	client_ctx.disconnect_data = "DISCONNECTED";
-	client_ctx.publish_data = "PUBLISH";
-
 	rc = network_setup();
 	PRINT_RESULT("network_setup", rc);
 	if (rc < 0) {
 		return;
 	}
 
-	rc = mqtt_init(&client_ctx.mqtt_ctx, MQTT_APP_PUBLISHER);
+	rc = mqtt_init();
 	PRINT_RESULT("mqtt_init", rc);
 	if (rc != 0) {
 		return;
 	}
 
-	for (i = 0; i < CONN_TRIES; i++) {
-		rc = mqtt_connect(&client_ctx.mqtt_ctx);
-		PRINT_RESULT("mqtt_connect", rc);
-		if (!rc) {
-			goto connected;
-		}
-	}
-
-	goto exit_app;
-
-connected:
-
 	rc = try_to_connect(&client_ctx);
 	PRINT_RESULT("try_to_connect", rc);
 	if (rc != 0) {
-		goto exit_app;
+		return;
 	}
 
 	i = 0;
-	while (i++ < APP_MAX_ITERATIONS) {
-		rc = mqtt_tx_pingreq(&client_ctx.mqtt_ctx);
-		k_sleep(APP_SLEEP_MSECS);
-		PRINT_RESULT("mqtt_tx_pingreq", rc);
-
-		prepare_mqtt_publish_msg(&client_ctx.pub_msg, MQTT_QoS0);
-		rc = mqtt_tx_publish(&client_ctx.mqtt_ctx, &client_ctx.pub_msg);
-		k_sleep(APP_SLEEP_MSECS);
-		PRINT_RESULT("mqtt_tx_publish", rc);
-
-		prepare_mqtt_publish_msg(&client_ctx.pub_msg, MQTT_QoS1);
-		rc = mqtt_tx_publish(&client_ctx.mqtt_ctx, &client_ctx.pub_msg);
-		k_sleep(APP_SLEEP_MSECS);
-		PRINT_RESULT("mqtt_tx_publish", rc);
-
-		prepare_mqtt_publish_msg(&client_ctx.pub_msg, MQTT_QoS2);
-		rc = mqtt_tx_publish(&client_ctx.mqtt_ctx, &client_ctx.pub_msg);
-		k_sleep(APP_SLEEP_MSECS);
-		PRINT_RESULT("mqtt_tx_publish", rc);
-	}
+	while (i++ < 4 * APP_MAX_ITERATIONS && connected) {
+		iter_start = k_uptime_get();
+
+		if (qos > MQTT_QoS_2_EACTLY_ONCE) {
+			rc = mqtt_ping(&client_ctx);
+			PRINT_RESULT("mqtt_ping", rc);
+		} else {
+			rc = publish(qos);
+			PRINT_RESULT("mqtt_publish", rc);
+		}
+
+		if (rc != 0) {
+			goto exit_app;
+		}
+
+		remaining = APP_SLEEP_MSECS;
+
+		while (remaining > 0 && connected) {
+			wait(remaining);
 
-	rc = mqtt_tx_disconnect(&client_ctx.mqtt_ctx);
-	PRINT_RESULT("mqtt_tx_disconnect", rc);
+			rc = mqtt_live();
+			if (rc != 0) {
+				PRINT_RESULT("mqtt_live", rc);
+				goto exit_app;
+			}
+
+			rc = mqtt_input(&client_ctx);
+			if (rc != 0) {
+				PRINT_RESULT("mqtt_input", rc);
+				goto exit_app;
+			}
+
+			remaining = APP_SLEEP_MSECS + iter_start -
+				    k_uptime_get();
+		}
+
+		qos++;
```

Comment:
Ah, so that's why there's a `>` operation against an enum value above. Well, worth a comment there.

And generally, relying on a particular enum value ordering/nearness may be a good optimization trick for internal library code, but looks somewhat hacky in a user-facing sample.